cmd/prune-junit-xml/prunexml.go

```diff
@@ -64,27 +66,68 @@ func main() {
 
 func pruneXML(suites *junitxml.JUnitTestSuites, maxBytes int) {
 	for _, suite := range suites.Suites {
-		for _, testcase := range suite.TestCases {
+		for i := range suite.TestCases {
+			// Modify directly in the TestCases slice, if necessary.
+			testcase := &suite.TestCases[i]
 			if testcase.SkipMessage != nil {
 				if len(testcase.SkipMessage.Message) > maxBytes {
 					fmt.Printf("clipping skip message in test case : %s\n", testcase.Name)
-					head := testcase.SkipMessage.Message[:maxBytes/2]
-					tail := testcase.SkipMessage.Message[len(testcase.SkipMessage.Message)-maxBytes/2:]
-					testcase.SkipMessage.Message = head + "[...clipped...]" + tail
+					pruneString(&testcase.SkipMessage.Message, maxBytes)
 				}
 			}
 			if testcase.Failure != nil {
+				// In Go unit tests, the entire test output
+				// becomes the failure message because `go
+				// test` doesn't track why a test fails. This
+				// can make the failure message pretty large.
+				//
+				// We cannot identify the real failure here
+				// either because Kubernetes has no convention
+				// for how to format test failures. What we can
+				// do is recognize log output added by klog.
+				//
+				// Therefore here we move the full text to
+				// to the test output and only keep those
+				// lines in the failure which are not from
+				// klog.
+				if testcase.SystemOut == "" {
+					if len(testcase.SystemOut) > maxBytes {
+						fmt.Printf("clipping log output in test case: %s\n", testcase.Name)
+						pruneString(&testcase.SystemOut, maxBytes)
+					}
+					var buf strings.Builder
+					haveLogOutput := false
+					for entry := range logparse.All(strings.NewReader(testcase.Failure.Contents)) {
+						if _, ok := entry.(*logparse.KlogEntry); ok {
+							haveLogOutput = true
+							continue
+						}
+						_, _ = buf.WriteString(entry.LogData())
+					}
+					if haveLogOutput {
+						// Store the full output
+						// because it is different from
+						// the updated failure message.
+						testcase.SystemOut = testcase.Failure.Contents
+					}
+					testcase.Failure.Contents = buf.String()
+				}
+
 				if len(testcase.Failure.Contents) > maxBytes {
 					fmt.Printf("clipping failure message in test case : %s\n", testcase.Name)
-					head := testcase.Failure.Contents[:maxBytes/2]
-					tail := testcase.Failure.Contents[len(testcase.Failure.Contents)-maxBytes/2:]
-					testcase.Failure.Contents = head + "[...clipped...]" + tail
+					pruneString(&testcase.Failure.Contents, maxBytes)
 				}
 			}
 		}
 	}
 }
 
+func pruneString(str *string, maxBytes int) {
```

Comment:
this seems it has a high risk of panicking, should we validate the string length?

Reply:
or just don't encapsulate in its own function , is not that is adding much value 3 lines vs 1 line

Reply:
or make it conditional as `pruneStringIfNeeded` that "prunes a string to the number of bytes passed as argument"

throwing some ideas, I'm worry in the future someone picks this and doesn't not validate the input

Reply:
Makes sense. I didn't want to move the message and its arguments into this helper because it then needs even more arguments, but it's better when doing the if check inside the helper. Changed.